asab/api/discovery.py

```diff
@@ -164,106 +171,113 @@ async def _rescan_advertised_instances(self):
+						elif ip == "::":
+							family = socket.AF_INET6
+						else:
+							continue
+
+						if discovery is not None:
+							for id_type, ids in discovery.items():
+								if advertised.get(id_type) is None:
+									advertised[id_type] = {}
+
+								for identifier in ids:
+									if identifier is not None:
+										if advertised[id_type].get(identifier) is None:
+											advertised[id_type][identifier] = {(host, port, family)}
+										else:
+											advertised[id_type][identifier].add((host, port, family))
+			except Exception:
+				L.exception("Error when scanning advertised instances")
+				return
 
 			self._advertised_cache = advertised
+			self._advertised_raw = advertised_raw
 			self._ready_event.set()
 
 
 	async def _iter_zk_items(self):
 		base_path = self.ZooKeeperContainer.Path + "/run"
 
 		def get_items():
+			result = []
 			try:
 				# Create the base path if it does not exist
 				if not self.ZooKeeperContainer.ZooKeeper.Client.exists(base_path):
 					self.ZooKeeperContainer.ZooKeeper.Client.create(base_path, b'', makepath=True)
 
-				return self.ZooKeeperContainer.ZooKeeper.Client.get_children(base_path, watch=self._on_change_threadsafe)
+				items = self.ZooKeeperContainer.ZooKeeper.Client.get_children(base_path, watch=self._on_change_threadsafe)
+
 			except (kazoo.exceptions.SessionExpiredError, kazoo.exceptions.ConnectionLoss):
 				L.warning("Connection to ZooKeeper lost. Discovery Service could not fetch up-to-date state of the cluster services.")
 				return None
 
-			except kazoo.exceptions.KazooException:
+			except Exception:
+				L.exception("Error when getting advertised instances")
 				return None
 
-		def get_data(item):
-			try:
-				data, stat = self.ZooKeeperContainer.ZooKeeper.Client.get((base_path + '/' + item), watch=self._on_change_threadsafe)
-				return data
-			except (kazoo.exceptions.SessionExpiredError, kazoo.exceptions.ConnectionLoss):
-				L.warning("Connection to ZooKeeper lost. Discovery Service could not fetch up-to-date state of the cluster services.")
-				return None
+			for item in items:
+				try:
+					data, stat = self.ZooKeeperContainer.ZooKeeper.Client.get(base_path + '/' + item, watch=self._on_change_threadsafe)
+					result.append((item, json.loads(data)))
+				except (kazoo.exceptions.SessionExpiredError, kazoo.exceptions.ConnectionLoss):
+					L.warning("Connection to ZooKeeper lost. Discovery Service could not fetch up-to-date state of the cluster services.")
+					return None
+				except kazoo.exceptions.NoNodeError:
+					continue
 
-			except kazoo.exceptions.KazooException:
-				# There's a race condition -> if ZK node is deleted between get_items and get_data calls, NoNodeError is raised. Let's just ignore it. The ZK node is already deleted.
-				return None
+			return result
 
-		items = await self.ProactorService.execute(get_items)
-		if items is None:
-			return
+		result = await self.ProactorService.execute(get_items)
+		if result is None:
```

Comment:
why did you remove the async generator? to make try/except easier?

im ok with that. the code is weird now... if None, return empty string... but ok.